### PR TITLE
Html to hiccup

### LINKED ISSUE #407: html->hiccup conversion
Heya,

This is a feature request.

Cursive, another popular editor for Clojure, has a handy feature where it will ask to convert HTML strings to Hiccup when pasting them into Clojure/Script files. This functionality is similar to what's provided in https://htmltohiccup.herokuapp.com/. It would be handy to have it in Calva as well.

### PR DESCRIPTION
**Pull request by PEZ**
## What has changed?

We add two commands:

* `calva.convertHtml2Hiccup`: In the same spirit as the converters for JS->CLJS and Dart->CLJS.
    * However, this commands takes arguments, you can provide the HTML to be converted and the conversion options
    * The command can be made to return the results instead of opening up an Untitled buffer
* `calva.pasteHtmlAsHiccup`: Also takes arguments, but only the conversion options, it will always operate on whatever is on the clipboard
    * Restores the clipboard after paste
    * Does not alter clipboard on conversion errors 

Example keyboard shortcuts:

```
    {
        "key": "ctrl+alt+c cmd+h",
        // Without args, the command uses the `calva.html2HiccupOptions` configuration
        // And writes the results to an Untitled document
        "command": "calva.convertHtml2Hiccup",
    },
    {
        "key": "ctrl+alt+c ctrl+h",
        "command": "calva.convertHtml2Hiccup",
        // With args, `"toUntitled": true` is necessary for keyboard shortcuts
        // without it, the command just returns the result to the caller
        "args": {"toUntitled": true, "options": {"kebab-attrs?": false}}
    },
    {
        "key": "ctrl+alt+c shift+h",
        "command": "calva.convertHtml2Hiccup",
        // Only for testing, providing the HTML is only useful from e.g. Joyride 
        "args": {"html": "<foo style='a: b' bar='baz'>gaz<foo>", "toUntitled": true}
    },
    {
        "key": "ctrl+alt+c h",
        // Without args, the command uses the `calva.html2HiccupOptions` configuration
        "command": "calva.pasteHtmlAsHiccup"
    },
    {
        "key": "ctrl+alt+c ctrl+alt+h",
        "command": "calva.pasteHtmlAsHiccup",
        // Override the `calva.html2HiccupOptions` configuration
        "args": {"mapify-style?": true, "kebab-attrs?": true}
    },
```

TODOs:

- [x] Add function to calva-lib for converting html->hiccup
- [x] Expose html->hiccup to calva-lib consumers
- [x] Create UI for converting HTML text to Hiccup
- [x] Document the commands and the API they create

* Fixes #407

Future:

I think the pattern here with commands accepting arguments is a good way to go with lots of Calva commands. As first step might be making the other converters take arguments. 

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
  - [x] Tested the particular change
  - [x] Figured if the change might have some side effects and tested those as well.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
  - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik

### CODE REVIEW

**Review by bpringe — APPROVED** [review]
Looking good to me. Fun stuff.

File: src/cljs-lib/test/calva/html2hiccup_test.cljs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    (is (= [[:foo [:bar]]]
+           (sut/html->hiccup "<foo><bar></bar></foo>"))))
+  (testing "Multiple top level elements are supported"
+    (is (= [[:foo] [:bar]]
+           (sut/html->hiccup "<foo></foo><bar></bar>"))))
+  (testing "Auto-closed tags are supported"
+    (is (= [[:foo]]
+           (sut/html->hiccup "<foo/>"))))
+  (testing "html comments `<!--...-->` becomes `(comment ...)`"
+    (is (= [[:foo nil]] ; TODO What's a way to actually test that it was a `(comment ...)`?
```
Comment (bpringe) [inline src/cljs-lib/test/calva/html2hiccup_test.cljs]: You have a failing test that looks like the function returns `(comment "...")`, but maybe this comment was added before that was happening?

Comment (bpringe) [reply src/cljs-lib/test/calva/html2hiccup_test.cljs]:
```
FAIL in (html->hiccup) (calva/html2hiccup_test.cljs:43:9)
html comments `<!--...-->` becomes `(comment ...)`
expected: (= [[:foo nil]] (sut/html->hiccup "<foo><!-- ... --></foo>"))
  actual: (not (= [[:foo nil]] [[:foo (comment "...")]]))
```

Comment (PEZ) [reply src/cljs-lib/test/calva/html2hiccup_test.cljs]: Yeah, I was being confused about how to test it. Now fixed. Thanks!

**Comment by PEZ** [comment]
Added some options.

**Comment by PEZ** [comment]
I need to add docs, but otherwise this is where I wanted to take this, @bpringe.

**Review by bpringe — APPROVED** [review]
Looks good. I think having some gifs of the usage of these commands in the docs would be good, or maybe a Calva TV video?

**Comment by PEZ** [comment]
I did a short video demonstrating the commands.

A narrated Calva TV vide would make sense standalone as well, I think.

**Pull request merged**